test/boost/cache_algorithm_test.cc

```diff
@@ -0,0 +1,152 @@
+//   because this results in the biggest `index page size`:`row size` ratio,
+//   due to details of the SSTable format.
+//   This makes the effects drastic, and so easy to test.
+// - The total size of all keys is at least 2 times greater than the size of RAM.
+//   This ensures that most of the index is uncached. This is necessary for
+//   the issue to become a visible.
+// - The size of user data (1000 per partition) is significantly bigger
+//   than several hundred bytes, to make various constant overheads
+//   (per-cell, per-row, per-partition) smaller than the size of user data.
+//   This simplifies reasoning about the test.
+//   In particular, it should ensure that each index page contains about 2000 keys,
+//   so it has size about 2 MiB.
+//
+// After populating this table, the test reads (sequentially) a subset of 1000 rows
+// multiple times. Since the total size of this hot subset (including overheads) is
+// only about 1 MiB, the test expects it to be perfectly cached.
+// This should be true unless index cache is flooding the cache.
+SEASTAR_TEST_CASE(test_index_doesnt_flood_cache_in_small_partition_workload) {
+    cql_test_config cfg;
+    cfg.db_config->task_ttl_seconds.set(0);
+    // As of this writing, uncommenting the below should make the test fail.
+    // cfg.db_config->index_cache_fraction.set(1.0);
+    return do_with_cql_env_thread([] (cql_test_env& e) {
+        // We disable compactions because they cause confusing cache mispopulations.
+        e.execute_cql("CREATE TABLE ks.t(pk blob PRIMARY KEY) WITH compaction = { 'class' : 'NullCompactionStrategy' };").get();
+        auto insert_query = e.prepare("INSERT INTO ks.t(pk) VALUES (?)").get0();
+        auto select_query = e.prepare("SELECT * FROM t WHERE pk = ?").get0();
+
+        constexpr uint64_t pk_number = 300000;
+        constexpr uint64_t pk_size = 1000;
+        // Sanity check. The test assumes that the total index size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(pk_size * pk_number, 2 * seastar::memory::stats().total_memory());
+
+        // A bijection between uint64_t and blobs of size pk_size.
+        auto make_key = [pk_size] (uint64_t x) {
+            bytes b(bytes::initialized_later(), std::max(pk_size, sizeof(x)));
+            auto i = b.begin();
+            write<uint64_t>(i, x);
+            return b;
+        };
+
+        // Populate the table.
+        for (size_t i = 0; i < pk_number; ++i) {
+            e.execute_prepared(insert_query, {{cql3::raw_value::make_value(make_key(i))}}).get0();
+        }
+        // Flushing makes reasoning easier.
+        e.db().invoke_on_all(&replica::database::flush_all_memtables).get();
+
+
+        constexpr uint64_t hot_subset_size = 1000;
+        // Sanity check. The test assumes that the total *hot* index size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(hot_subset_size * pk_size * 2000, 2 * seastar::memory::stats().total_memory());
```

Comment:
2000 is the `data file size : summary file size` ratio. Hence, in a table containing only primary keys, `pk_size * 2000` is approximately the size of an index page.

The sanity check here is that the maximum total size of the touched index pages is much greater than RAM. ("Maximum", because multiple rows can land in the same index page).

I'll add a comment which explains that.